src/Microsoft.SqlTools.ServiceLayer/FileBrowser/FileBrowserService.cs

```diff
@@ -0,0 +1,222 @@
+            FileBrowserOpenParams fileBrowserParams,
+            RequestContext<FileBrowserOpenResponse> requestContext)
+        {
+            FileBrowserOpenResponse response = new FileBrowserOpenResponse();
+
+            try
+            {
+                ConnectionInfo connInfo;
+                this.ConnectionServiceInstance.TryFindConnection(
+                    fileBrowserParams.OwnerUri,
+                    out connInfo);
+
+                if (connInfo != null)
+                {
+                    SqlConnection sqlConn = GetSqlConnection(connInfo);
+                    FileBrowserOperation browser = new FileBrowserOperation(sqlConn, fileBrowserParams.ExpandPath, fileBrowserParams.FileFilters);
+                    browser.PopulateFileTree();
+
+                    // Cache the file browser operation for expanding node request in the future
+                    this.ownerToFileBrowserMap.Add(fileBrowserParams.OwnerUri, browser);
+
+                    response.FileTree = browser.FileTree;
+                    response.Succeeded = true;
+                }
+            }
+            catch (Exception ex)
+            {
+                response.Succeeded = false;
+                response.Message = ex.Message;
+            }
+
+            await requestContext.SendResult(response);
+        }
+
+        internal async Task HandleFileBrowserExpandRequest(
+            FileBrowserExpandParams fileBrowserParams,
+            RequestContext<FileBrowserExpandResponse> requestContext)
+        {
+            FileBrowserExpandResponse response = new FileBrowserExpandResponse();
+
+            try
+            {
+                FileBrowserOperation browser = this.ownerToFileBrowserMap[fileBrowserParams.OwnerUri];
+                if (browser != null)
+                {
+                    browser.ExpandSelectedNode(fileBrowserParams.ExpandPath);
+                    response.ExpandedNode = browser.FileTree.SelectedNode;
+                    response.Succeeded = true;
+                }
+            }
+            catch (Exception ex)
+            {
+                response.Succeeded = false;
+                response.Message = ex.Message;
+            }
+
+            await requestContext.SendResult(response);
+        }
+
+        internal async Task HandleFileBrowserValidateRequest(
+            FileBrowserValidateParams fileBrowserParams,
+            RequestContext<FileBrowserValidateResponse> requestContext)
+        {
+            FileBrowserValidateResponse response = new FileBrowserValidateResponse();
+
+            try
+            {
+                if (fileBrowserParams.SelectedFiles != null && fileBrowserParams.SelectedFiles.Length > 0)
+                {
+                    IFileValidator fileValidator = null;
+                    switch (fileBrowserParams.ServiceType)
+                    {
+                        case ServiceConstants.Backup:
+                        case ServiceConstants.Restore:
```

Comment:
it might be a more flexible design to have services register validator callbacks rather than hard-coding the logic for other services here?

Reply:
changed